src/charm.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm Jenkins."""
+
+import logging
+from time import sleep
+from typing import TYPE_CHECKING, Any
+
+import requests
+from jenkinsapi.jenkins import Jenkins
+from ops.charm import CharmBase, PebbleReadyEvent
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus
+from ops.pebble import Layer
+
+from path import INITIAL_PASSWORD, JENKINS_HOME, LAST_EXEC, UPDATE_VERSION
+from types_ import Credentials
+
+if TYPE_CHECKING:
+    from ops.pebble import LayerDict  # pragma: no cover
+
+logger = logging.getLogger(__name__)
+
+
+class JenkinsK8SOperatorCharm(CharmBase):
+    """Charm Jenkins."""
+
+    def __init__(self, *args: Any):
+        """Initialize the charm and register event handlers.
+
+        Args:
+            args: Arguments to initialize the char base.
+        """
+        super().__init__(*args)
+        self.framework.observe(self.on.jenkins_pebble_ready, self._on_jenkins_pebble_ready)
+
+    def _get_pebble_layer(self, env: dict[str, str] | None = None) -> Layer:
+        """Return a dictionary representing a Pebble layer.
+
+        Args:
+            env: Map of Jenkins environment variables.
+
+        Returns:
+            The pebble layer defining Jenkins service layer.
+        """
+        default_env = {"JENKINS_HOME": str(JENKINS_HOME)}
+        merged_env = default_env | env if env else default_env
+        layer: LayerDict = {
+            "summary": "jenkins layer",
+            "description": "pebble config layer for jenkins",
+            "services": {
+                "jenkins": {
+                    "override": "replace",
+                    "summary": "jenkins",
+                    "command": "java -Djava.awt.headless=true -jar /srv/jenkins/jenkins.war",
+                    "startup": "enabled",
+                    "environment": merged_env,
+                }
+            },
+        }
+        return Layer(layer)
+
+    def _is_jenkins_ready(self) -> bool:
+        """Check if Jenkins webserver is ready.
+
+        Returns:
+            True if Jenkins server is online. False otherwise.
+        """
+        try:
```

Comment:
How about `return requests.get("http://localhost:8080/login", timeout=10).ok` instead?

Reply:
This is great! Thank you :D